README.md
# Prepper

Prepper is a simple server provisioning tool, built on top of SSHKit. You can
use it to script your server build process. It is heavily inspired by (Sprinkle)[https://github.com/sprinkle-tool/sprinkle], but Sprinkle doesn't seem to be maintained anymore, that's why Prepper was born.


## Installation

    $ gem install prepper

## Usage

Prepper works with "packages". You define a package with a name and pass it a block.
Within that block you can execute commands on the target host.
There are built in helpers to install `apt` packages, manage directories and
upload files to the server, etc.

A simple example:

```ruby

server_host "YOUR_SERVER_IP"
server_port 22
server_user "root"

# let's install the necessary packages to run a Rails app with Postgresql
package :apt do
  apt_update
  apt_install %w(git-core build-essential libcurl4 libcurl4-openssl-dev libjemalloc-dev postgresql-client libpq-dev postgresql-contrib)
end

# now we will add a deploy user
package :add_deploy_user do
  add_user 'deploy', shell: '/bin/bash', flags: '--disabled-password'

  directory '/home/deploy/.ssh', owner: 'deploy:deploy'
  file '/home/deploy/.ssh/authorized_keys', owner: 'deploy:deploy', mode: '655', content: 'ssh-rsa YOUR PUBLIC SSH KEY'
  file '/etc/sudoers.d/deploy', owner: 'root:root', template: 'sudoers'
end

# install rbenv and Ruby 3.1.2
package :install_ruby do
  install_rbenv 'deploy'
  install_ruby 'deploy', '3.1.2', '--with-jemalloc'
end

```

You can see a full example in [examples/config.rb](examples/config.rb). You would run that file with `bundle exec prepper config.rb` to provision the server.

The API documentation can be found on rubydoc: [https://rubydoc.info/github/gregmolnar/prepper](https://rubydoc.info/github/gregmolnar/prepper).

## Development

After checking out the repo, run `bin/setup` to install dependencies. Then, run `rake test` to run the tests. You can also run `bin/console` for an interactive prompt that will allow you to experiment.

To install this gem onto your local machine, run `bundle exec rake install`. To release a new version, update the version number in `version.rb`, and then run `bundle exec rake release`, which will create a git tag for the version, push git commits and tags, and push the `.gem` file to [rubygems.org](https://rubygems.org).

## Contributing

Bug reports and pull requests are welcome on GitHub at https://github.com/gregmolnar/prepper. This project is intended to be a safe, welcoming space for collaboration, and contributors are expected to adhere to the [code of conduct](https://github.com/gregmolnar/prepper/blob/master/CODE_OF_CONDUCT.md).


## License

The gem is available as open source under the terms of the [MIT License](https://opensource.org/licenses/MIT).

## Code of Conduct

Everyone interacting in the Prepper project's codebases, issue trackers, chat rooms and mailing lists is expected to follow the [code of conduct](https://github.com/gregmolnar/prepper/blob/master/CODE_OF_CONDUCT.md).


CHANGELOG.md
# Change log

## Unreleased

## 0.2.0

* better docs

## 0.1.0

* First release


examples/config.rb
server_host "YOUR_SERVER_IP"
server_port 22
server_user "root"

# let's install the necessary packages to run a Rails app with Postgresql
package :apt do
  apt_update
  apt_install %w(git-core build-essential libcurl4 libcurl4-openssl-dev libjemalloc-dev postgresql-client libpq-dev postgresql-contrib)
end

# now we will add a deploy user
package :add_deploy_user do
  add_user 'deploy', shell: '/bin/bash', flags: '--disabled-password'

  directory '/home/deploy/.ssh', owner: 'deploy:deploy'
  file '/home/deploy/.ssh/authorized_keys', owner: 'deploy:deploy', mode: '655', content: 'ssh-rsa YOUR PUBLIC SSH KEY'
  file '/etc/sudoers.d/deploy', owner: 'root:root', template: 'sudoers'
end

# install rbenv and Ruby 3.1.2
package :install_ruby do
  install_rbenv 'deploy'
  install_ruby 'deploy', '3.1.2', '--with-jemalloc'
end

# install yarn
package :yarn do
  add_command 'curl -sS https://dl.yarnpkg.com/debian/pubkey.gpg | sudo apt-key add -'
  add_command 'echo "deb https://dl.yarnpkg.com/debian/ stable main" | sudo tee /etc/apt/sources.list.d/yarn.list',
    verify: has_file?('/etc/apt/sources.list.d/yarn.list')
  apt_update
  apt_install %w(nodejs yarn)
end

# install the caddy webserver
package :install_caddy do
  apt_install %w(debian-keyring debian-archive-keyring apt-transport-https)
  add_command "curl -1sLf 'https://dl.cloudsmith.io/public/caddy/stable/gpg.key' | sudo gpg --dearmor -o /usr/share/keyrings/caddy-stable-archive-keyring.gpg", verifier: has_file?('/etc/apt/sources.list.d/caddy-stable.list')
  add_command "curl -1sLf 'https://dl.cloudsmith.io/public/caddy/stable/debian.deb.txt' | sudo tee /etc/apt/sources.list.d/caddy-stable.list", verifier: has_file?('/etc/apt/sources.list.d/caddy-stable.list')

  apt_update
  apt_install %w(caddy)
  directory "/etc/caddy/sites", owner: 'caddy:caddy'
  file "/etc/caddy/Caddyfile", content: "import /etc/caddy/sites/*.caddy", owner: 'caddy:caddy'
  file "/etc/caddy/sites/global.caddy", content: "
    {
      debug
      log {
        output file /var/log/caddy/caddy.log {
          roll_size 10MB
        }
      }
    }
  ", owner: 'caddy:caddy'
  chown '/etc/caddy/*', 'caddy:caddy'
  add_command 'sudo adduser caddy deploy'
  add_command 'sudo service caddy reload'
end

# create a static site
package :my_blog_com do
  directory '/home/deploy/domains/myblog.com/public'
  chown '/home/deploy/domains/myblog.com', 'deploy:deploy', "-R"
  directory '/home/deploy/domains/myblog.com/shared/log', owner: "caddy:caddy"
  file '/etc/caddy/sites/myblog.com.caddy', template: 'myblog.com.caddy'
  chown '/etc/caddy/sites', 'caddy:caddy', "-R"
  add_command "sudo service caddy reload"
end

# create a vhost for a Rails site
package :myrails_site_com do
  directory '/home/deploy/domains/myrails-site.com/'
  chown '/home/deploy/domains/', 'deploy:deploy'
  file '/etc/caddy/sites/myrails-site.com.caddy', template: 'myrails-site.com.caddy'
  chown '/etc/caddy/sites/*', 'caddy:caddy'
  add_command "sudo service caddy reload"
end

# create a systemd service for puma
package :puma_myrails_site do
  directory "/home/deploy/.config/systemd/user/", user: 'deploy'
  chown "/home/deploy/.config", "deploy:deploy", "-R"
  file "/home/deploy/.config/systemd/user/puma_myrails-site.service", owner: 'deploy:deploy', template: 'puma_myrails-site.service'

  directory '/home/deploy/.config/systemd/user/default.target.wants', user: 'deploy'
  symlink "/home/deploy/.config/systemd/user/default.target.wants/puma_myrails-site.service", "/home/deploy/.config/systemd/user/puma_myrails-site.service", user: 'deploy'

  add_command "sudo -u deploy -l systemctl --user daemon-reload"
  add_command "sudo -u deploy -l systemctl --user enable puma_myrails-site"
  add_command "sudo loginctl enable-linger deploy"
end
